config/functions

```diff
@@ -478,33 +478,46 @@ debug_strip() {
 
 init_package_cache() {
   local _ANCHOR="@?+?@" DIR
+  local temp_global temp_local
 
   # If the package caches are unset, then populate them
   if [ -z "${_CACHE_PACKAGE_LOCAL}" -o -z "${_CACHE_PACKAGE_GLOBAL}" ]; then
-    _CACHE_PACKAGE_LOCAL="${BUILD}/.cache_package_local"
-    _CACHE_PACKAGE_GLOBAL="${BUILD}/.cache_package_global"
-    mkdir -p "${BUILD}"
-    : > "${_CACHE_PACKAGE_LOCAL}"
-    : > "${_CACHE_PACKAGE_GLOBAL}"
+    temp_global="$(mktemp)"
+    temp_local="$(mktemp)"
 
     # cache project/device folder for a package
     if [ -n $DEVICE ]; then
       for DIR in $(find $ROOT/projects/$PROJECT/devices/$DEVICE/packages -type d 2>/dev/null); do
-        [ -r "$DIR/package.mk" ] && echo "${DIR}${_ANCHOR}" >> "${_CACHE_PACKAGE_LOCAL}"
+        [ -r "$DIR/package.mk" ] && echo "${DIR}${_ANCHOR}" >> "${temp_local}"
       done
     fi
 
     # cache project folder for a package
     for DIR in $(find $ROOT/projects/$PROJECT/packages -type d 2>/dev/null); do
-      [ -r "$DIR/package.mk" ] && echo "${DIR}${_ANCHOR}" >> "${_CACHE_PACKAGE_LOCAL}"
+      [ -r "$DIR/package.mk" ] && echo "${DIR}${_ANCHOR}" >> "${temp_local}"
     done
 
     # cache packages folder
     for DIR in $(find $ROOT/$PACKAGES -type d 2>/dev/null); do
-      [ -r "$DIR/package.mk" ] && echo "${DIR}${_ANCHOR}" >> "${_CACHE_PACKAGE_GLOBAL}"
+      [ -r "$DIR/package.mk" ] && echo "${DIR}${_ANCHOR}" >> "${temp_global}"
     done
 
+    _CACHE_PACKAGE_LOCAL="${BUILD}/.cache_package_local"
+    _CACHE_PACKAGE_GLOBAL="${BUILD}/.cache_package_global"
     export _CACHE_PACKAGE_LOCAL _CACHE_PACKAGE_GLOBAL
+
+    # overwrite existing cache files only when they are invalid, or not yet created
+    mkdir -p "$(dirname "${_CACHE_PACKAGE_GLOBAL}")"
+    if [[ ! -f "${_CACHE_PACKAGE_LOCAL}" ]] || [[ "$(cat "${temp_local}" | sha256sum)" != "$(cat "${_CACHE_PACKAGE_LOCAL}" | md5sum)" ]]; then
```

Comment:
I don't understand this check - you seem to comparing the sha256 sum of the temp_local file with the md5sum of the cache file?
Also any reason why you use [[ instead of [?

Reply:
> Also any reason why you use [[ instead of [?
Yes, `[[` will short-circuit, while `[` does not. So if the file doesn't exist, we don't attempt to compare hashes (which also means we don't need to discard the error that would be generated by reading a non-existent `$BUILD/.cache_package_*`).

Reply:
Short circuiting is already done by `||`. And instead of calculating hashes we can compare the files directly - that makes the check a bit simpler. eg
```
if [ -f "${_CACHE_PACKAGE_LOCAL}" ] && cmp -s "${temp_local}" "${_CACHE_PACKAGE_LOCAL}" ; then
  rm "${temp_local}"
else
  mv "${temp_local}" "${_CACHE_PACKAGE_LOCAL}"
fi
```
The file existence check could be dropped, cmp will give an exit code of 2 if the file doesn't exist, but IMO the explicit check version is more readable

Reply:
> Short circuiting is already done by ||.

You are correct, thanks. I've updated with your suggestions.